app/forms/firewall-rules-common.tsx

```diff
@@ -26,95 +37,277 @@ import * as MiniTable from '~/ui/lib/MiniTable'
+  >
+    <MiniTable.Header>
+      <MiniTable.HeadCell>Type</MiniTable.HeadCell>
+      <MiniTable.HeadCell>Value</MiniTable.HeadCell>
+      {/* For remove button */}
+      <MiniTable.HeadCell className="w-12" />
+    </MiniTable.Header>
+    <MiniTable.Body>
+      {items.value.map(({ type, value }, index) => (
+        <MiniTable.Row
+          tabIndex={0}
+          aria-rowindex={index + 1}
+          aria-label={`Name: ${value}, Type: ${type}`}
+          key={`${type}|${value}`}
         >
-          API docs
-        </a>{' '}
-        to learn more about firewall rules.
-      </>
-    }
-  />
+          <MiniTable.Cell>
+            <Badge variant="solid">{type}</Badge>
+          </MiniTable.Cell>
+          <MiniTable.Cell>{value}</MiniTable.Cell>
+          <MiniTable.RemoveCell
+            onClick={() =>
+              items.onChange(
+                items.value.filter((i) => !(i.value === value && i.type === type))
+              )
+            }
+            label={`remove ${sectionType} ${value}`}
+          />
+        </MiniTable.Row>
+      ))}
+    </MiniTable.Body>
+  </MiniTable.Table>
+)
+
+// Given an array of committed items (VPCs, Subnets, Instances) and
+// a list of all items, return the items that are available
+const availableItems = (
+  committedItems: Array<VpcFirewallRuleTarget | VpcFirewallRuleHostFilter>,
+  itemType: 'vpc' | 'subnet' | 'instance',
+  items: Array<Vpc | VpcSubnet | Instance>
+) => {
+  if (!items) return []
+  return (
+    items
+      .map((item) => item.name)
+      // remove any items that match the committed items on both type and value
+      .filter(
+        (name) =>
+          !committedItems.filter((ci) => ci.type === itemType && ci.value === name).length
+      )
+      .map((name) => ({ label: name, value: name }))
+  )
+}
+
+type ProtocolFieldProps = {
+  control: Control<FirewallRuleValues>
+  protocol: 'TCP' | 'UDP' | 'ICMP'
+}
+const ProtocolField = ({ control, protocol }: ProtocolFieldProps) => (
+  <div>
+    <CheckboxField name="protocols" value={protocol} control={control}>
+      {protocol}
+    </CheckboxField>
+  </div>
 )
 
-export const CommonFields = ({ error, control, nameTaken }: CommonFieldsProps) => {
-  const portRangeForm = useForm({ defaultValues: portRangeDefaultValues })
+type CommonFieldsProps = {
+  control: Control<FirewallRuleValues>
+  nameTaken: (name: string) => boolean
+  error: ApiError | null
+}
+
+export const CommonFields = ({ control, nameTaken, error }: CommonFieldsProps) => {
+  const { project, vpc } = useVpcSelector()
+  const targetAndHostDefaultValues: TargetAndHostFormValues = {
+    type: 'vpc',
+    value: '',
+    // only becomes relevant when the type is 'VPC subnet'; ignored otherwise
+    subnetVpc: vpc,
+  }
+  // prefetchedApiQueries below are prefetched in firewall-rules-create and -edit
+  const {
+    data: { items: instances },
+  } = usePrefetchedApiQuery('instanceList', {
+    query: { project, limit: PAGE_SIZE },
```

Comment:
One last thing I thought of: I think instance and VPC list calls should either have no `limit`, which uses a [default value of 100](https://github.com/oxidecomputer/dropshot/blob/d3218d30634dc6419a00a75d56ed98891d0f3c03/dropshot/src/server.rs#L138), or maybe they should have an explicit page size of like 1000, which we use in a few spots when we mean "approximately everything": https://github.com/search?q=repo%3Aoxidecomputer%2Fconsole%20%22limit%3A%201000%22&type=code

Make sure to change both the prefetches and these hook calls if you change it.